internal/stacks/stackruntime/internal/stackeval/main_apply.go

```diff
@@ -117,10 +132,66 @@ func ApplyPlan(ctx context.Context, config *stackconfig.Config, rawPlan []*anypb
 								"Inconsistent component instance plan",
 								fmt.Sprintf("The plan for %s is inconsistent with its prior state: %s.", addr, err),
 							))
+							log.Printf("[ERROR] stackeval: %s has a plan inconsistent with its prior state: %s", addr, err)
+							span.SetStatus(codes.Error, "plan is inconsistent with prior state")
 							return nil, diags
 						}
 
-						return inst.ApplyModuleTreePlan(ctx, modulesRuntimePlan)
+						var waitForComponents collections.Set[stackaddrs.AbsComponent]
+						if action == plans.Delete {
+							// If the effect of this apply will be to destroy this
+							// component instance then we need to wait for all
+							// of our dependents to be destroyed first, because
+							// we're required to outlive them.
+							//
+							// (We can assume that all of the dependents are
+							// also performing destroy plans, because we'd have
+							// rejected the configuration as invalid if a
+							// downstream component were referring to a
+							// component that's been removed from the config.)
+							waitForComponents = dependentAddrs
+						} else {
+							// For all other actions, we must wait for our
+							// dependencies to finish applying their changes.
+							waitForComponents = dependencyAddrs
+						}
+						if depCount := waitForComponents.Len(); depCount != 0 {
+							log.Printf("[TRACE] stackeval: %s waiting for its predecessors (%d) to complete", addr, depCount)
+						}
+						for _, waitComponentAddr := range waitForComponents.Elems() {
+							if stack := main.Stack(ctx, waitComponentAddr.Stack, ApplyPhase); stack != nil {
+								if component := stack.Component(ctx, waitComponentAddr.Item); component != nil {
+									span.AddEvent("awaiting predecessor", trace.WithAttributes(
+										attribute.String("component_addr", waitComponentAddr.String()),
+									))
+									success := component.ApplySuccessful(ctx)
+									if !success {
+										// If anything we're waiting on does not succeed then we can't proceed without
+										// violating the dependency invariants.
+										log.Printf("[TRACE] stackeval: %s cannot start because %s changes did not apply completely", addr, waitComponentAddr)
+										span.AddEvent("predecessor is incomplete", trace.WithAttributes(
+											attribute.String("component_addr", waitComponentAddr.String()),
+										))
+										span.SetStatus(codes.Error, "predecessors did not completely apply")
+
+										// We'll return a stub result that reports that nothing was changed, since
+										// we're not going to run our apply phase at all.
+										return inst.PlaceholderApplyResultForSkippedApply(ctx, modulesRuntimePlan), nil
+										// We also skip emitting any progress events for this component instance,
+										// since it didn't even _start_ its apply process.
```

Comment:
I was referring here to the entire torrent of events that get emitted while `inst.ApplyModuleTreePlan` is doing its work, since we're not running that function at all and so there won't be any events about applying this component _at all_.

I agree though that it isn't really clear what this comment is referring to. When I was writing it I was aiming for drawing attention to the fact that we don't emit progress events in this case without making a comment here refer too much to the details of an entirely-separate function in a different file, since I worry that future maintenance would make a more specific comment invalid and confusing.

However, I also can't argue that this comment isn't _already_ confusing. Do you think that an explicit mention of "the events from inst.ApplyModuleTreePlan" would be sufficient to connect the dots here, assuming a reader who is already at least a little familiar with how Terraform Core and Terraform Cloud Agent are collaborating here?

Reply:
Ah, I see. I think the cause of my confusion was wording and the specific scope of "we". If I understand the situation correctly, this might have been clearer to me:

> Note that skipping the apply process also means that Terraform will not emit any progress events for this component instance at all.

I think your suggestion would work too!

Reply:
I will blend these two ideas together. :grinning: